### PR TITLE
Feat/add pam

### PR DESCRIPTION
**Pull request by envin3**

### CODE REVIEW

File: cpp/contracts/adapter.cpp [hunk 1]
```diff
@@ -46,16 +46,21 @@ void adapter::create(
 
    auto _token_bytes = token_bytes.extract_as_byte_array();
    check(_token_bytes.size() == 32, "token bytes length must be 32");
-   check(is_account(token), "token account does not exist");
+
+   // Do not check token validity if token is not EOS native
+   if (token != name(0)) check(is_account(token), "token account does not exist");
```
Comment (gitmp01) [inline cpp/contracts/adapter.cpp:51]: Be careful that there's no native distinction on EOS: it's a token as the other ones, check the stats table on [eosio.token](https://bloks.io/account/eosio.token?loadContract=true&tab=Tables&table=stat&account=eosio.token&scope=EOS&limit=100)

In other words, if EOS is being wrapped `token` would be equal to `eosio.token`, which is a valid account

File: cpp/contracts/adapter.cpp [hunk 2]
```diff
@@ -46,16 +46,21 @@ void adapter::create(
...
    check(is_account(xerc20), "xERC20 account does not exist");
-   check(token_symbol.precision() == xerc20_symbol.precision(), "invalid xerc20 precision");
+
+   // Difference in precision allowed if token is not EOS native
+   if (token != name(0)) check(token_symbol.precision() == xerc20_symbol.precision(), "invalid xerc20 precision");
```
Comment (gitmp01) [inline cpp/contracts/adapter.cpp:55]: Same thing here and other similar lines

**Review by gitmp01 — COMMENTED** [review]
I was starting the review but I stopped because it seems you are still working on it (settee isn't putting the attestation in the table). So I left a comment in the meantime, the I'll review when the code is ready

File: cpp/contracts/adapter.cpp (hunk 1 above)
Comment (envin3) [reply cpp/contracts/adapter.cpp:51]: Here I mean, do not check if the token is not native in a  EOS chain like. With name(0), which is address = '' I expect that the native token is from some other chain type (like EVM for exampel)

Comment (gitmp01) [reply cpp/contracts/adapter.cpp:51]: Ah you're right, I missed that, sorry! 

The "native" word in the comment misled me, I think it's better to stick to our definition though, I started to call it "local" referring to the unwrapped version chain, because "native" goes in conflict with the xerc20 definition of native, which is the chain's currency

File: cpp/contracts/adapter.cpp [hunk 3]
```diff
@@ -147,28 +152,104 @@ void adapter::freeuserdata(const name& account) {
...
    require_auth(caller);
 
    registry_adapter _registry(get_self(), get_self().value);
    auto idx_registry = _registry.get_index<adapter_registry_idx_token_bytes>();
    auto search_token_bytes = idx_registry.find(operation.token);
-
    check(search_token_bytes != idx_registry.end(), "invalid token");
 
-   checksum256 event_id;
-   check(pam::is_authorized(operation, metadata, event_id), "unauthorized");
```
Comment (gitmp01) [inline cpp/contracts/adapter.cpp:160]: Why removing this? If it was for get_self() not being available within `is_authorized`, you can always pass it as arg like this
```cpp
pam::is_authorized(get_self(), operation, metadata, event_id);
```

Comment (gitmp01) [reply cpp/contracts/adapter.cpp:160]: Then the tables should be accessible

File: cpp/contracts/adapter.cpp [hunk 4]
```diff
@@ -147,28 +152,104 @@ void adapter::freeuserdata(const name& account) {
...
+   tee_pubkey _tee_pubkey(get_self(), get_self().value);
+   public_key tee_key = _tee_pubkey.get().key;
+
+   uint128_t a = 2;
+   bytes origin_chain_id = pam::extract_32bytes(metadata.preimage, a);
+
+   mappings_table _mappings_table(get_self(), get_self().value);
+
+   auto itr_mappings = _mappings_table.find(pam::get_mappings_key(origin_chain_id));
+   check(itr_mappings != _mappings_table.end(), "Unauthorized: origin chain_id not registered");
```
Comment (gitmp01) [inline cpp/contracts/adapter.cpp:238]: I've seen in general that the check message is lowercase, you can also remove the unauthorized is would come out as
```eosio_assert: unauthorized: ... ```

File: cpp/contracts/adapter.cpp [hunk 5]
```diff
@@ -147,28 +152,104 @@ void adapter::freeuserdata(const name& account) {
...
+   bytes exp_emitter = itr_mappings->emitter;
+   bytes exp_topic_zero =  itr_mappings->topic_zero;
+   pam::check_authorization(operation, metadata, event_id, tee_key, exp_emitter, exp_topic_zero);
```
Comment (gitmp01) [inline cpp/contracts/adapter.cpp:241]: Yeah I would push to bring everything into the `is_authorized` method if possible

File: cpp/contracts/adapter.cpp [hunk 6]
```diff
@@ -334,5 +416,174 @@ void adapter::onmint(const name& caller, const name& to, const asset& quantity,
...
+
+signature pam::convert_bytes_to_signature(const bytes& input_bytes) {
+   check(input_bytes.size() == 65, "Signature must be exactly 65 bytes");
+   std::array<char, 65> sig_data;
+   std::copy(input_bytes.begin(), input_bytes.end(), sig_data.begin());
+   return signature(std::in_place_index<0>, sig_data);
+}
+
+bool pam::context_checks(const operation& operation, const metadata& metadata) {
+   uint8_t offset = 2; // Skip protocol, verion
```
Comment (gitmp01) [inline cpp/contracts/adapter.cpp:432]:
```suggestion
   uint8_t offset = 2; // Skip protocol, version
```

File: cpp/contracts/adapter.cpp [hunk 7]
```diff
@@ -334,5 +416,174 @@ void adapter::onmint(const name& caller, const name& to, const asset& quantity,
...
+}
+
+checksum256 pam::bytes32_to_checksum256(const bytes& data) {
+   check(data.size() == 32, "The input must be 32 bytes long.");
+   std::array<uint8_t, 32> byte_array;
+   std::copy(data.begin(), data.end(), byte_array.begin());
+   return checksum256(byte_array);
+}
+
+name pam::bytes_to_name(const bytes& data) {
```
Comment (gitmp01) [inline cpp/contracts/adapter.cpp:513]: I'd move all these conversion helpers inside `utils.hpp`

File: cpp/test/adapter-eth-eos.test.js [hunk 8]
```diff
@@ -0,0 +1,281 @@
...
+
+  const TABLE_STORAGE = 'storage'
+  const FEE_BASIS_POINTS = 1750
+  const FEE_BASIS_POINTS_DIVISOR = 1000000
+
+  const token = {
+    symbol: symbol,
+    account: `${symbol.toLowerCase()}.token`,
+    maxSupply: `${maxSupply} ${symbol}`,
+    bytes: '0000000000000000000000003ca5269b5c54d4c807ca0df7eeb2cb7a5327e77d', //tokenBytes,
```
Comment (gitmp01) [inline cpp/test/adapter-eth-eos.test.js:67]: Easier for the reader
```suggestion
    bytes: '0000000000000000000000003ca5269b5c54d4c807ca0df7eeb2cb7a5327e77d', // eth address bytes representation
```

File: cpp/test/adapter-eth-eos.test.js
Comment (gitmp01) [inline cpp/test/adapter-eth-eos.test.js:1]: Generalize the names to something like `adapter-local.test.js` and the other one to `adapter-non-local.test.js`

**Review by gitmp01 — COMMENTED** [review]
Few comments to sort out, we'll definitely need to re-iterate on tests since we need to make a proper suite reduced to the bare minimum in order to test just the pam authorization, this for easy maintenance

File: cpp/contracts/pam.hpp [hunk 9]
```diff
@@ -1,14 +1,50 @@
...
-         return true;
-      }
+    using bytes = std::vector<uint8_t>;
+    namespace pam {
+        TABLE mappings {
+            bytes chain_id;
+            bytes emitter;
+            bytes topic_zero;
+
+            EOSLIB_SERIALIZE(mappings, (chain_id)(topic_zero)(emitter));
```
[the remark this answers is not in the document]
Comment (envin3) [reply cpp/contracts/pam.hpp]: It can be removed.

File: cpp/test/utils/utils.js
[the remark this answers is not in the document]
Comment (gitmp01) [reply cpp/test/utils/utils.js:1]: A note: if you wanted to get an easy way to import the utils functions it shouldn't be by including everything into the same .js module. 

The nice way to it it's by declaring an index.js file inside the utils/ folder and export all the module functions from there

```js
// index.js

const getEventBytes = require('./get-event-bytes')
// ...

module.exports = {
  ...getEventBytes,
  // ...
}
```

Then the import would be like

```js
const { getEventBytes } = require('./utils')
```

**Review by gitmp01 — CHANGES_REQUESTED** [review]

File: cpp/contracts/pam.hpp [hunk 10]
```diff
@@ -32,8 +30,8 @@ namespace eosio {
             public_key key;
         };
 
-        // typedef eosio::singleton<"tee"_n, tee> tee_pubkey;
-        // typedef eosio::multi_index<"mappings"_n, mappings> mappings_table;
+        using tee_pubkey = singleton<"tee"_n, pam::tee>;
+        using lockbox_singleton = singleton<"lockbox"_n, name>;
```
[the remark this answers is not in the document]
Comment (envin3) [reply cpp/contracts/pam.hpp:34]: refactoring flood, restored.

File: cpp/contracts/xerc20.token.cpp [hunk 11]
```diff
@@ -35,8 +35,9 @@ void xtoken::mint( const name& caller, const name& to, const asset& quantity, co
     check( existing != statstable.end(), "token with symbol does not exist, create token before issue" );
     const auto& st = *existing;
 
+    name lockbox;
     lockbox_singleton _lockbox( get_self(), get_self().value );
-    auto lockbox = _lockbox.get();
+    if (_lockbox.exists()) lockbox = _lockbox.get();
```
[the remark this answers is not in the document]
Comment (envin3) [reply cpp/contracts/xerc20.token.cpp:40]: Interesting. I didn't know about the #46 before, but yes it solves that.

File: cpp/contracts/adapter.cpp [hunk 12]
```diff
@@ -334,5 +402,79 @@ void adapter::onmint(const name& caller, const name& to, const asset& quantity,
...
+   name recipient_name = bytes_to_name(recipient);
+   check(operation.recipient == recipient_name, "recipient do not match");
+   offset += recipient_len_num;
+
+   //check user data -- FIXME --
+   bytes user_data(event_data.begin() + offset, event_data.end());
+   // TODO fix user data decoding
+   checksum256 data256 = sha256((const char*)user_data.data(), user_data.size());
+   checksum256 op_data256 = sha256((const char*)operation.data.data(), operation.data.size());
+}
```
[the remark this answers is not in the document]
Comment (envin3) [reply cpp/contracts/adapter.cpp:478]: good catch
